src/util/include/util/glib_casts.h

```diff
@@ -0,0 +1,120 @@
+
+namespace xoj::util {
+
+namespace detail {
+
+template <typename F>
+struct FunctionTraits;
+
+template <typename R, typename... Args>
+struct FunctionTraits<R (&)(Args...)> {
+    using Reference = R (&)(Args...);
+    using RetType = R;
+    using ArgTypes = std::tuple<Args...>;
+    static constexpr std::size_t ArgCount = sizeof...(Args);
+    template <std::size_t N>
+    using NthArg = std::tuple_element_t<N, ArgTypes>;
+    using FirstArg = NthArg<0>;
+    using LastArg = NthArg<ArgCount - 1>;
+};
+
+template <typename R, typename... Args>
+struct FunctionTraits<R (*)(Args...)> {
+    using Pointer = R (*)(Args...);
+    using RetType = R;
+    using ArgTypes = std::tuple<Args...>;
+    static constexpr std::size_t ArgCount = sizeof...(Args);
+    template <std::size_t N>
+    using NthArg = std::tuple_element_t<N, ArgTypes>;
+    using FirstArg = NthArg<0>;
+    using LastArg = NthArg<ArgCount - 1>;
+};
+
+template <typename F, size_t... Is>
+constexpr auto indices_impl(F f, std::index_sequence<Is...>) {
+    return f(std::integral_constant<size_t, Is>()...);
+}
+
+template <size_t N, typename F>
+constexpr auto indices(F f) {
+    return indices_impl(f, std::make_index_sequence<N>());
+}
+
+template <auto SrcFn, class ST, class... Args>
+constexpr auto callback_once_impl(Args... args, void* data) -> gboolean {
+    SrcFn(args..., static_cast<ST>(data));
+    return G_SOURCE_REMOVE;
+}
+
+template <auto SrcFn, class R, class ST, class... Args>
+constexpr auto wrap_impl(Args... args, void* data) -> std::conditional_t<std::is_same_v<bool, R>, gboolean, R> {
+    if constexpr (std::is_same_v<void, R>) {
+        SrcFn(args..., static_cast<ST>(data));
+        return;
+    } else if constexpr (std::is_same_v<bool, R>) {
+        return SrcFn(args..., static_cast<ST>(data)) ? G_SOURCE_CONTINUE : G_SOURCE_REMOVE;
+    } else if (!std::is_same_v<void, R>) {
+        return SrcFn(args..., static_cast<ST>(data));
+    } else
+        static_assert(!std::is_same_v<R, void>, "Invalid return type");
+}
+
+}  // namespace detail
+
+template <auto SrcFn>
+constexpr auto wrap_for_once() -> auto {
+    using namespace detail;
+    using FTI = FunctionTraits<decltype(SrcFn)>;
+    return indices<FTI::ArgCount - 1>([&](auto... Is) {
+        return callback_once_impl<
+                SrcFn, typename FTI::LastArg,
+                std::remove_reference_t<decltype(std::get<Is>(std::declval<typename FTI::ArgTypes>()))>...>;
+    });
+}
+
+template <auto SrcFn>
+constexpr auto wrap() -> auto {
+    using namespace detail;
+    using FTI = FunctionTraits<decltype(SrcFn)>;
+    return indices<FTI::ArgCount - 1>([&](auto... Is) {
+        return wrap_impl<SrcFn, typename FTI::RetType, typename FTI::LastArg,
+                         std::remove_reference_t<decltype(std::get<Is>(std::declval<typename FTI::ArgTypes>()))>...>;
+    });
+}
+
+template <auto SrcFn>
+constexpr auto wrap_for_g_callback() -> GCallback {
+    return G_CALLBACK(wrap<SrcFn>());
+}
```

Comment:
Would consider add some docs on these wrappers? Specially about when to use them.

Reply:
Added some documentation.